### ISSUE TITLE
Jupyter Notebook Autoscaling issue in Kubeflow

### ISSUE THREAD

**Issue by SheharyarTkxel** [kubernetes/autoscaler#2273]
**Background**
I have my Cluster Autoscaler all set in Kubeflow. It works fine. I tested it by deploying NGINX deployments with multiple replicas to overwhelm my GPU based AWS Servers to scale up/down them. However, I was expecting the same behavior in Kubeflow's Jupyter Notebook.

**What I tried**
What I did was write some Deep Learning code with Keras library and tried to overwhelm the servers by setting a high learning rate. It did not scale up the nodes. In fact it didn't even reach out to CA to scale up as observed in the Cluster Autoscaler log. I realize that an ML problem in Jupyter Notebook might not be capable of Distributed Computing since the Jupyter Notebook pod is deployed on a single node and is contained entirely in that node as far as I understand.

**What I want**
Let me know if a code in Jupyter Notebook is even capable of distributing in terms of scaling nodes up/down to solve a problem that demands more computational power on the go? If yes then how can I configure the notebook to make use of Cluster Autoscaler to scale up/down?

**Comment by Denton24646** [comment]
The notebook pod itself cannot use the CA (Cluster Autoscaler) directly - it likely doesn't have the permissions to make new pods that would trigger the scheduler and then the CA into spinning up new nodes for the new pods. You should use a higher level of abstraction to create multiple pods, seems like you can use a `TFJob` to run distributed training for your notebook.

> When you bundle Jupyter Notebooks in Kubeflow, you can use the Fairing library to submit training jobs using TFJob. The training job can run single node or distributed on the same Kubernetes cluster, but not inside the notebook pod itself. Submitting the job with the Fairing library makes processes like Docker containerization and pod allocation clear for data scientists.

https://www.kubeflow.org/docs/notebooks/why-use-jupyter-notebook/

This seems to turn your notebook into a static docker image and then creates a job based on that image, with one or more pods. This is something like the behavior you're describing. 

I think really what you're describing is the ability to dynamically scale up and down node resources based on the GPU utilization of the node itself - if the learning rate is high and the GPU memory is low scale up another node. This is interesting but hard - to go from a singlenode to multinode training on the fly is pretty difficult - lots of networking and shared storage needs to be provisioned and allocated dynamically. 

You can setup the horizontal pod autoscaler to spin up new pods based on custom metrics like GPU memory usage being over a certain threshold. Use pod affinity/anti-affinity to ensure that the new pods cannot be scheduled to existing notebook nodes. This would in effect cause the CA to spin up new nodes for the newly created pods. Then in your code you would have to have the ability to add new workers to the training job as its running.

**Comment by Jeffwan** [comment]
All the code running inside Jupyter Notebook will use resource of that pod.. In order to leverage CA capability to scale up/down your cluster. You want to make sure your training/serving pods running outside Jupyter Notebook..

**Comment by SheharyarTkxel** [comment]
> The notebook pod itself cannot use the CA (Cluster Autoscaler) directly - it likely doesn't have the permissions to make new pods that would trigger the scheduler and then the CA into spinning up new nodes for the new pods. You should use a higher level of abstraction to create multiple pods, seems like you can use a `TFJob` to run distributed training for your notebook.
> 
> > When you bundle Jupyter Notebooks in Kubeflow, you can use the Fairing library to submit training jobs using TFJob. The training job can run single node or distributed on the same Kubernetes cluster, but not inside the notebook pod itself. Submitting the job with the Fairing library makes processes like Docker containerization and pod allocation clear for data scientists.
> 
> https://www.kubeflow.org/docs/notebooks/why-use-jupyter-notebook/
> 
> This seems to turn your notebook into a static docker image and then creates a job based on that image, with one or more pods. This is something like the behavior you're describing.
> 
> I think really what you're describing is the ability to dynamically scale up and down node resources based on the GPU utilization of the node itself - if the learning rate is high and the GPU memory is low scale up another node. This is interesting but hard - to go from a singlenode to multinode training on the fly is pretty difficult - lots of networking and shared storage needs to be provisioned and allocated dynamically.
> 
> You can setup the horizontal pod autoscaler to spin up new pods based on custom metrics like GPU memory usage being over a certain threshold. Use pod affinity/anti-affinity to ensure that the new pods cannot be scheduled to existing notebook nodes. This would in effect cause the CA to spin up new nodes for the newly created pods. Then in your code you would have to have the ability to add new workers to the training job as its running.

@Denton24646 Thanks. Makes sense. I've started following this [tutorial](https://www.kubeflow.org/docs/fairing/fairing-overview/) on Fairing. Hopefully it'll provide some solution.
